drivers/storage/portworx/portworx_test.go

```diff
@@ -981,14 +981,18 @@ func TestStorageClusterDefaultsForCSI(t *testing.T) {
 			Namespace: "kube-test",
 		},
 		Spec: corev1.StorageClusterSpec{
-			Image: "px/image:2.1.5.1",
+			Image: "px/image:2.9.0.1",
 		},
 	}
+	// Simulate DesiredImages.CSISnapshotController being empty for old operator version w/o this image
+	driver.SetDefaultsOnStorageCluster(cluster)
+	cluster.Status.DesiredImages.CSISnapshotController = ""
```

Comment:
After the above change we will have to split this test into two steps. 
First, just enable CSI - DesiredImages.CSISnapshotController should be empty
Second, enable snapshot controller - DesiredImages.CSISnapshotController should not be empty

Reply:
Made the above changes.

SetDefaultsOnStorageCluster seems to still always set the `desiredImages.SnapshotController`. I had to keep this line to clear it:
```
cluster.Status.DesiredImages.CSISnapshotController = ""
```

However, when SetDefaultsOnStorageCluster is called again with InstallSnapshotController=true set, the image is set properly. This is expected, right?

Reply:
Actually, it is not expected, but I guess it does not have a bad effect. So far we have populated the desired images of only components that are enabled. Let's try to be consistent just to be safe.
I see what the issue is after comparing to other components. Our `autoUpdateCSISnapshotController` is a bit incorrect.
```
func autoUpdateCSISnapshotController(cluster *corev1.StorageCluster) bool {
	return cluster.Spec.CSI.InstallSnapshotController != nil && *cluster.Spec.CSI.InstallSnapshotController
}

func hasCSIChanged(cluster *corev1.StorageCluster) bool {
	return pxutil.IsCSIEnabled(cluster) &&
		(cluster.Status.DesiredImages.CSIProvisioner == "" || hasCSISnapshotControllerChanged(cluster))
}

func hasCSISnapshotControllerChanged(cluster *corev1.StorageCluster) bool {
         return autoUpdateCSISnapshotController(cluster) && 
                      cluster.Status.DesiredImages.CSISnapshotController == ""

}
```
and line 249 should be - 
```
			if autoUpdateCSISnapshotController(toUpdate) &&
                                 (cluster.Status.DesiredImages.CSISnapshotController == "" ||
				 pxVersionChanged ||
				 autoUpdateComponents(toUpdate) {
```

These changes should be inline with other components and should work as expected.

---
Another issue I observed is that we are not resetting the snapshot controller desired image when it is disabled. We should add this after line 315 - 
```
        if !pxutil.IsCSIEnabled(toUpdate) {
              ...
	else if !autoUpdateCSISnapshotController(toUpdate) {
		toUpdate.Status.DesiredImages.CSISnapshotController = ""
	}
```

Reply:
Makes sense. Just made those changes. Test passes now without the image reset line:
```
cluster.Status.DesiredImages.CSISnapshotController = ""
```

Reply:
Could we add a test to just disable csi snapshot controller and check that the desired image for controller is reset?